### PR TITLE
bug: allow negative NodeIDs for magic variables

### PR DESCRIPTION
**Pull request by moodlezoup**
Addresses #1

### CODE REVIEW

File: tests/ast-json/Counter.json [hunk 1]
```diff
@@ -1,244 +1,407 @@
...
-      "id": 1,
-      "literals": [
-        "solidity",
-        "^",
-        "0.8",
-        ".13"
-      ],
-      "nodeType": "PragmaDirective",
-      "src": "39:24:0"
+    "absolutePath": "Counter.sol",
```
Comment (moodlezoup) [inline tests/ast-json/Counter.json:2]: Added some magic variable stuff:

```
// SPDX-License-Identifier: UNLICENSED
pragma solidity ^0.8.13;

contract Counter {
    uint256 public number;

    function setNumber(uint256 newNumber) public {
        keccak256(abi.encode(address(this).balance));
        number = newNumber;
    }

    function increment() public {
        number++;
    }
}
```

**Comment by hrkrshnn** [comment]
Thanks for the PR, I'll review this in a couple of days. Initially I thought it was just a matter of changing `pub type NodeId = i64`, but looks like I've to carefully examine here!

**Comment by moodlezoup** [comment]
> Thanks for the PR, I'll review this in a couple of days. Initially I thought it was just a matter of changing `pub type NodeId = i64`, but looks like I've to carefully examine here!

yeah that would work too, but I tried to make it a bit more safe

File: src/ast/node.rs [hunk 2]
```diff
@@ -1,6 +1,39 @@
...
+    Suicide = -24,
+    Super = -25,
+    Tx = -26,
+    Type = -27,
+    This = -28,
+}
+
+#[derive(Clone, Copy, Debug, Deserialize, Serialize, PartialEq, Eq)]
+#[serde(untagged)]
+pub enum NodeID {
```
Comment (hrkrshnn) [inline src/ast/node.rs]: I think I'll go with `i64` to have parity with the Solidity's represenation (https://github.com/ethereum/solidity/blob/3f2cde9bd6073d27bab709d5759a4f55bcbeb006/libsolidity/ast/AST.h#L82)

File: tests/ast-json/Counter.json [hunk 3]
```diff
@@ -1,244 +1,407 @@
...
+                                        "name": "keccak256",
+                                        "nodeType": "Identifier",
+                                        "overloadedDeclarations": [],
+                                        "referencedDeclaration": -8,
+                                        "src": "171:9:17",
+                                        "typeDescriptions": {
+                                            "typeIdentifier": "t_function_keccak256_pure$_t_bytes_memory_ptr_$returns$_t_bytes32_$",
+                                            "typeString": "function (bytes memory) pure returns (bytes32)"
+                                        }
+                                    },
```
Comment (hrkrshnn) [inline tests/ast-json/Counter.json:207]: Note: here the `id` is actually an unsigned number. The enum you created there is used to represent `referencedDeclarations`.

File: src/ast/node.rs [hunk 4]
```diff
@@ -1,6 +1,39 @@
...
+    Ripemd160 = -20,
+    Selfdestruct = -21,
+    Sha256 = -22,
+    Sha3 = -23,
+    Suicide = -24,
+    Super = -25,
+    Tx = -26,
+    Type = -27,
+    This = -28,
+}
```
Comment (hrkrshnn) [inline src/ast/node.rs]: This is still a valuable addition. I'll make a new file `magic.rs` and put this there. Thanks

**Review by hrkrshnn — COMMENTED** [review]
Thanks for this. I'll just push the changes now and merge this!

**Comment by hrkrshnn** [comment]
Once again, thank you @moodlezoup for spotting this. Also curious what you are planning to build with this. Feel free to DM me on twitter (_hrkrshnn) if you have ideas on using the library.

**Pull request merged**